### PR TITLE
Optimize IsOverridden logic in DynamicObject

### PR DESCRIPTION
**Pull request by bartdesmet**
The `IsOverridden` method in `DynamicObject` looks quite baroque and uses a method name to re-retrieve `MethodInfo` objects for the `Try*` methods defined on `DynamicObject`. It seems we can do a much better job by passing the `MethodInfo` object to the method.

I'm marking this as WIP because we don't have good test coverage in this area. @JonHanna, are you currently working on tests in this area? If not, I can help out to cover the gaps related to `DynamicObject`.

I've also left a `TODO` with some observation on the logic of the code. It seems to allow binding to methods in the subclass that have the same name but a different return type. I'd have expected it to check for a proper virtual override of the base definition. We can write tests to assert all the current behaviors and then evaluate what to do here (if anything).

### CODE REVIEW

**Comment by JonHanna** [comment]
I am. I'm intending to get that folder up toward 100% coverage and then look at dynamic expressions, which are also 0% covered. I'm planning to do the remaining binders and then circle back on DynamicObject after that, since anything exercising it is likely to make use of the binders and it's nice when green lines in coverage reports show where you've actually done testing and it's off your todo list rather than showing code that got passed through testing something else.

I think the behaviour you refer to makes sense in terms of reflecting how method hiding happens, for some value of "makes sense"; in strongly-typed C# when there is hiding we can opt for a base or derived invocation by which type we call through, but since with dynamic calls its the instance at run-time rather than the type at compile time that decides if the binding will work, the author had to either decide to never respect hiding or never let it be by-passed. Since in a way late-binding will always work with the full set of methods no matter where in an inheritance hierarchy they were declared, in a way late binding is akin to always dealing with the object's actual type and never a base, and so always accepting hiding is IMO the most reasonable option in a dilemma where neither choice is perfect.

**Comment by bartdesmet** [comment]
@dotnet-bot test Innerloop Ubuntu14.04 Debug Build and Test please (https://github.com/dotnet/corefx/issues/13979)

**Comment by bartdesmet** [comment]
The behavior of `IsOverridden` differs from the original at https://referencesource.microsoft.com/#System.Core/Microsoft/Scripting/Actions/DynamicObject.cs,42ff7d1d8461d3f2. It seems to me this change got introduced when `GetBaseDefinition` got missing in the APIs before it came back.

**Comment by bartdesmet** [comment]
See https://github.com/dotnet/corefx/pull/14116 for a suggested change to bring back the original `IsOverridden`. If accepted, the latest revision of this PR builds on this by adding some other improvements to the way `IsOverridden` is used.

File: src/System.Linq.Expressions/src/System/Dynamic/DynamicObject.cs [hunk 1]
```diff
@@ -227,44 +227,64 @@ internal MetaDynamic(Expression expression, DynamicObject value)
 
             public override DynamicMetaObject BindGetMember(GetMemberBinder binder)
             {
-                if (IsOverridden(nameof(DynamicObject.TryGetMember)))
+                if (IsOverridden(DynamicObject_TryGetMember))
                 {
-                    return CallMethodWithResult(nameof(DynamicObject.TryGetMember), binder, s_noArgs, (e) => binder.FallbackGetMember(this, e));
+                    return BindGetMemberOverridden(binder);
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Dynamic/DynamicObject.cs]: Avoid unnecessary closures.

**Comment by bartdesmet** [comment]
The remaining changes in this PR (after merging in https://github.com/dotnet/corefx/pull/14122 and https://github.com/dotnet/corefx/pull/14116) avoid re-retrieving methods from the `DynamicObject` base class using `GetMethod`. Caching these now and passing `MethodInfo` objects rather than the `string` representing their name.

**Comment by stephentoub** [comment]
@dotnet-bot test this please

File: src/System.Linq.Expressions/src/System/Dynamic/DynamicObject.cs [hunk 2]
```diff
@@ -816,13 +816,13 @@ private DynamicMetaObject CallMethodNoResult<TBinder>(string methodName, TBinder
...
-            private bool IsOverridden(string method)
+            private bool IsOverridden(MethodInfo method)
             {
-                MemberInfo[] methods = Value.GetType().GetMember(method, MemberTypes.Method, BindingFlags.Public | BindingFlags.Instance);
+                MemberInfo[] methods = Value.GetType().GetMember(method.Name, MemberTypes.Method, BindingFlags.Public | BindingFlags.Instance);
 
                 foreach (MethodInfo mi in methods)
                 {
-                    if (mi.DeclaringType != typeof(DynamicObject) && mi.GetBaseDefinition().DeclaringType == typeof(DynamicObject))
+                    if (mi.DeclaringType != typeof(DynamicObject) && mi.GetBaseDefinition() == method)
```
Comment (JonHanna) [inline src/System.Linq.Expressions/src/System/Dynamic/DynamicObject.cs]: NM. It's not cheap in the implementations that come up here.

**Comment by VSadov** [comment]
@dotnet-bot test this please

**Comment by VSadov** [comment]
PR is relatively old and Mac tests seem to be stuck. Just rerun all tests.

**Comment by VSadov** [comment]
LGTM. Thanks.

**Pull request merged**